api/Updated-Metrics.md

```diff
@@ -0,0 +1,346 @@
+A = pd.DataFrame(np.transpose([A_sex, A_race]), columns=['Sex', 'Race'])
+
+weights = [ 1, 2, 3, 2, 2, 1, ...]
+```
+
+We actually seek to be very agnostic as to the contents of the `y_true` and `y_pred` arrays.
+Meaning is imposed on them by the underlying metrics.
+
+## Basic Calls
+
+### Existing Syntax
+
+Our basic method is `group_summary()`
+
+```python
+>>> result = flm.group_summary(skm.accuracy_score, y_true, y_pred, sensitive_features=A_sex)
+>>> print(result)
+{'overall': 0.4, 'by_group': {'male': 0.6536, 'female': 0.213}}
+>>> print(type(result))
+<class 'sklearn.utils.Bunch'>
+```
+The `Bunch` is an object which can be accessed in two ways - either as a dictionary - `result['overall']` -  or via properties named by the dictionary keys - `result.overall`.
+Note that the `by_group` key accesses another `Bunch`.
+
+We allow for sample weights (and other arguments which require slicing) via `indexed_params`, and passing through other arguments to the underlying metric function (in this case, `normalize`):
+```python
+>>> flm.group_summary(skm.accuracy_score, y_true, y_pred, sensitive_features=A_sex, indexed_params=['sample_weight'], sample_weight=weights, normalize=False)
+{'overall': 20, 'by_group': {'male': 60, 'female': 21}}
+```
+
+We also provide some wrappers for common metrics from SciKit-Learn:
+```python
+>>> flm.accuracy_score_group_summary(y_true, y_pred, sensitive_features=A_sex)
+{'overall': 0.4, 'by_group': {'male': 0.6536, 'female': 0.213}}
+```
+
+### Proposed Change
+
+We do not intend to change the API invoked by the user.
+What will change is the return type.
+Rather than a `Bunch`, we will return a `GroupedMetric` object, which can offer richer functionality.
+
+At this basic level, there is only a slight change to the results seen by the user.
+There are still properties `overall` and `by_groups`, with the same semantics.
+However, the `by_groups` result is now a Pandas Series, and we also provide a `metric` property to record the name of the underlying metric:
+```python
+>>> result = flm.group_summary(skm.accuracy_score, y_true, y_pred, sensitive_features=A_sex)
+>>> result.metric
+"sklearn.metrics.accuracy_score"
+>>> result.overall
+0.4
+>>> result.by_groups
+Male      0.6536
+Female    0.2130
+Name: sklearn.metrics.accuracy_score dtype: float64
+>>> print(type(result.by_groups))
+<class 'pandas.core.series.Series'>
+```
+The `metric` property (and using it as the name of the Series) could prove troublesome.
+This is because the fully qualified function name, as reconstructed from `__name__`, `__qualname` and `__module__` might not match the user's expectation.
+For example:
+```python
+>>> import sklearn.metrics as skm
+>>> skm.accuracy_score.__name__
+'accuracy_score'
+>>> skm.accuracy_score.__qualname__
+'accuracy_score'
+>>> skm.accuracy_score.__module__
+'sklearn.metrics._classification'
+```
+We are seeing some of the actual internal structure here of SciKit-Learn, and the user might not be expecting that.
```

Comment:
@romanlutz makes the same point about linking to the metric function itself above. I can see the logic in having a `metric_function_` property on the object as you suggest. I do see the names as being particularly useful for the multiple-metric case, though. Otherwise, the only way to find out what each column is, is to remember what order they specified the metric functions in the call to `group_summary()`

Reply:
Changed as you suggest, but the problem of naming the series remains.

Reply:
One way of solving the naming issue would be to allow for dicts, i.e. something like `{'accuracy' : sklearn.metrics.accuracy_score}`. But this might make things unnecessarily complex.